### PR TITLE
Implement per subsystem message capacity override

### LINKED ISSUE #24: Implement per subsystem message_capacity override.
Something like `#[subsystem(message_capacity=32768, SomeMessage, sends: [])]`.  

This is needed to properly size the channels for subsystems that need to handle occasionally large bursts of work without blocking any potential sender.

### PR DESCRIPTION
**Pull request by vstakhov**
Resolves #24

### CODE REVIEW

File: orchestra/proc-macro/src/parse/parse_orchestra_struct.rs [hunk 1]
```diff
@@ -56,8 +59,10 @@ impl Parse for SubSysAttrItem {
 			Self::Wip(input.parse::<kw::wip>()?)
 		} else if lookahead.peek(kw::blocking) {
 			Self::Blocking(input.parse::<kw::blocking>()?)
-		} else if lookahead.peek(kw::sends) {
+		} else if lookahead.peek(kw::consumes) {
```
Comment (drahnr) [inline orchestra/proc-macro/src/parse/parse_orchestra_struct.rs:62]: Was this intentional?

**Review by sandreim — APPROVED** [review]
LGTM, can we also add a test ?

**Comment by sandreim** [comment]
> Addresses #24

Would be nice to properly link issue https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue

File: orchestra/proc-macro/src/parse/parse_orchestra_struct.rs (hunk 1 above)
Comment (vstakhov) [reply orchestra/proc-macro/src/parse/parse_orchestra_struct.rs:62]: Yes, it is an occasional IDE damage, but this is a draft PR, so it is not yet completed :)

File: orchestra/proc-macro/src/parse/parse_orchestra_struct.rs [hunk 2]
```diff
@@ -289,6 +299,25 @@ impl Parse for Consumes {
...
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct MessageCapacity {
+	#[allow(dead_code)]
+	tag: kw::message_capacity,
+	#[allow(dead_code)]
+	eq_token: Token![=],
+	value: usize,
+}
```
[the remark this answers is not in the document]
Comment (vstakhov) [reply orchestra/proc-macro/src/parse/parse_orchestra_struct.rs:309]: I have concluded the same, as the parsing logic is quite different.

**Comment by vstakhov** [comment]
I have one doubt: for now, if we manually override channel size by calling the specific method in the builder, then it takes priority over per-subsystem sizes. Maybe, it is better to take `max(subsystem_specific_capacity, override_capacity)`? What do you think, @drahnr, @sandreim ?

**Comment by sandreim** [comment]
> I have one doubt: for now, if we manually override channel size by calling the specific method in the builder, then it takes priority over per-subsystem sizes. Maybe, it is better to take `max(subsystem_specific_capacity, override_capacity)`? What do you think, @drahnr, @sandreim ?

I am in favor for the per subsystem override to have priority over the high level value

**Comment by vstakhov** [comment]
> > I have one doubt: for now, if we manually override channel size by calling the specific method in the builder, then it takes priority over per-subsystem sizes. Maybe, it is better to take `max(subsystem_specific_capacity, override_capacity)`? What do you think, @drahnr, @sandreim ?
> 
> I am in favor for the per subsystem override to have priority over the high level value

It makes overrides confusing. So we can have 3 overrides: per overseer (lowest priority), per subsystem, per builder method. I think that the former two should have the same priority and the maximum value should be taken.

**Comment by sandreim** [comment]
> > > I have one doubt: for now, if we manually override channel size by calling the specific method in the builder, then it takes priority over per-subsystem sizes. Maybe, it is better to take `max(subsystem_specific_capacity, override_capacity)`? What do you think, @drahnr, @sandreim ?
> > 
> > 
> > I am in favor for the per subsystem override to have priority over the high level value
> 
> It makes overrides confusing. So we can have 3 overrides: per overseer (lowest priority), per subsystem, per builder method. I think that the former two should have the same priority and the maximum value should be taken.

I am confused a bit, I was expecting this is only available on the overseer and the subsystem. Wouldn't the builder refer to the overseer value ?

**Review by drahnr — APPROVED** [review]

**Comment by drahnr** [comment]
Generally I'd prefer it to stay declarative, but the builder pattern already has the same for the per orchestra override, so fine by me

**Pull request merged**